File: 0005-longest-palindromic-substring/0005-longest-palindromic-substring.py
class Solution:
    def longestPalindrome(self, s: str) -> str:
        ans = ""

        for i in range(len(s)):
            substr1 = self.getLongest(s,i,i)
            substr2 = self.getLongest(s, i, i + 1)
            if len(substr1) > len(ans):
                ans = substr1
            if len(substr2) > len(ans):
                ans = substr2
        
        return ans
    
    def getLongest(self, s, i, j):
        ans = 0
        while i >= 0 and j <= len(s) - 1 and s[i] == s[j]:
            i -= 1
            j += 1
        return s[i + 1: j]

File: 0005-longest-palindromic-substring/0005-longest-palindromic-substring.cpp
class Solution {
public:
    string longestPalindrome(string s) {
        vector<vector<bool>> dp(s.size(), vector<bool>(s.size(), 0));

        for (int len = 1; len <= s.size(); len++) {
            int j = len - 1;
            for (int i = 0; i < s.size() - len + 1; i++) {
                if (len == 1) dp[i][j] = true;
                else if (len == 2) dp[i][j] = s[i] == s[j];
                else {
                    dp[i][j] = s[i] == s[j] and dp[i + 1][j - 1];
                }
                j++;
            }
        }

        string res = s.substr(0, 1);
        int val = 1;


        for (int i = 0; i < s.size(); i++) {
            for (int j = i + 1; j < s.size(); j++) {
                if (dp[i][j]) {
                    if (j - i + 1 > val) {
                        val = j - i + 1;
                        res = s.substr(i, j - i + 1);
                    }
                } 
            }
        }
        return res;
    }
};